### PR TITLE
feat: provide repository contributors insights

### LINKED ISSUE #5: Feature: provide contributor insights
### Type of feature

🍕 Feature

### Current behavior

Currently, via the CLI, there's no way to see insights for a provided repos.

### Suggested solution

We should add an `insights` command that has a `contributors` subcommand to start.

When invoked with a git repo, it provides some very basic insights based on the commits to the user:

```
$ pizza insights contributors <git-repo>
Over last 30 days

New: 6
Recurring: 2
```

Should also provide the ability to specify the time range:
```
$ pizza insights contributors <git-repo> --time 60d

Over last 60 days:

New contributors: 6
Recurring contributors: 2
```

If the repo is a GitHub repo (which we have indexed data for from the API backend), we should generate a link to the insights page:

```
$ pizza insights contributors https://github.com/open-sauced/pizza-cli --time 60d

Over last 60 days:

New contributors: 6
Recurring contributors: 2

See more at: https://insights.opensauced.pizza/{insight-page}
```

The above will include a few calls to the API so there are some changes that probably need to be made there as well.

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [X] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by cecobask**
## Description
This pull request adds the feature to provide repository contributors insights.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
Closes #5

### File contents:
```
$ cat repos.yaml

- https://github.com/kubernetes/kubernetes
- https://github.com/cli/cli
```

### Usage with `--file` / `-f` flag
```
$ pizza insights contributors --file repos.yaml

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/cli/cli ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 25                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 30                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 40                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 5                          ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/kubernetes/kubernetes ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 158                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 280                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 197                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 122                                      ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

### Usage with `--repos` / `-r` flag
```
$ pizza insights contributors --repos "https://github.com/kubernetes/kubernetes,https://github.com/cli/cli"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/cli/cli ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 25                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 30                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 40                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 5                          ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/kubernetes/kubernetes ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 158                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 280                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 197                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 122                                      ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

### Ignore non-existent and non-indexed repositories
```
$ pizza insights contributors --repos "https://github.com/fake-owner/fake-repo,https://github.com/kubernetes/kubernetes"

ignoring repository issue: repository https://github.com/fake-owner/fake-repo is either non-existent or has not been indexed yet
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/kubernetes/kubernetes ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 158                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 280                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 197                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 122                                      ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

### Error when neither `--file` or `--repos` flag was provided
```
$ pizza insights contributors

Error: must pass either --file or --repos flag
Usage:
  pizza insights contributors [flags]

Flags:
  -f, --file string     Path to yaml file containing an array of git repository urls
  -h, --help            help for contributors
  -p, --period int      Number of days, used for query filtering (default 30)
  -r, --repos strings   Comma-separated array of git repository urls

Global Flags:
      --beta              Shorthand for using the beta OpenSauced API endpoint ("https://beta.api.opensauced.pizza/v1"). supercedes the '--endpoint' flag
  -e, --endpoint string   The API endpoint to send requests to (default "https://api.opensauced.pizza/v1")

2023/08/20 19:36:16 must pass either --file or --repos flag
```

### Error when both `--file` and `--repos` flags were provided
```
$ pizza insights contributors --file repos.yaml --repos "https://github.com/kubernetes/kubernetes,https://github.com/cli/cli"

Error: if any flags in the group [file repos] are set none of the others can be; [file repos] were all set
Usage:
  pizza insights contributors [flags]

Flags:
  -f, --file string     Path to yaml file containing an array of git repository urls
  -h, --help            help for contributors
  -p, --period int      Number of days, used for query filtering (default 30)
  -r, --repos strings   Comma-separated array of git repository urls

Global Flags:
      --beta              Shorthand for using the beta OpenSauced API endpoint ("https://beta.api.opensauced.pizza/v1"). supercedes the '--endpoint' flag
  -e, --endpoint string   The API endpoint to send requests to (default "https://api.opensauced.pizza/v1")

2023/08/20 19:36:42 if any flags in the group [file repos] are set none of the others can be; [file repos] were all set
```

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [x] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Review by jpmcb — CHANGES_REQUESTED** [review]
I'm realizing this has quite abit of overlap with #24.

We'll want to align the user experience so that there is a clear story for CLI users with each command

This will also be blocked on us getting an API client for Go: https://github.com/open-sauced/pizza-cli/issues/28

File: pkg/api/constants.go
Comment (brandonroberts) [reply pkg/api/constants.go:4]: Not often. We used it before when prototyping against different infrastructure, but haven't needed to use it since then. I don't think we need to include it

File: cmd/insights/contributors.go
Comment (k1nho) [inline cmd/insights/contributors.go:214]: Very cool concurrency going on here! have been looking into also doing this on #24 as well to parallelize the data collection of the contributors so that the TUI can get the initial model started faster. Some questions if you don't mind. For this part, could it be the case that this is a dead goroutine when an error is returned from one of the data collectors, like the select is blocking once the error is acquired from ```errorChan``` closes the channel and then returns nil and err but waitGroup is not done, so ```doneChan``` does not get close?

Also,  could this be a good case to use [errGroup](https://pkg.go.dev/golang.org/x/sync@v0.3.0/errgroup)? @jpmcb

Comment (cecobask) [reply cmd/insights/contributors.go:231]: Hi @k1nho, thanks!
The `waitGroup` will always close, regardless if an error occurred or not. This is because the start of each go routine explicitly defers the call to `waitGroup.Done()`.
In my opinion, this is a suitable place to use `errGroup`. It just comes down to style/preference. Behind the hood, they're doing something similar if I'm not wrong.

**Comment by bdougie** [comment]
> I'm realizing this has quite abit of overlap with #24.
> 
> We'll want to align the user experience so that there is a clear story for CLI users with each command

I see the difference being the interactivity you get from the show. I didn't want to increase scope on #24 while we explore the basic interaction, but think we can do some more high level insights from commits and prs as well.

**Comment by jpmcb** [comment]
@cecobask - the Go API client has landed ✨ https://github.com/open-sauced/go-api !!!! Are you able to give that a try? It likely has some sharp edges so please open issues in that repo as the need arises!

**Comment by jpmcb** [comment]
Can you change this PR's target branch to be `beta`? 👀 We recently changed the default branch to `beta` in support of using the releaser action

For further reading:
- https://github.com/open-sauced/pizza-cli/issues/34

**Comment by cecobask** [comment]
@jpmcb, Done 👍🏻 Another review session is in order 😄 
- Refactored the codebase to make API calls using the go client where possible, the only exception being `auth` and `repo-query` commands
- Consolidated duplicated logic and constants
- Configured `gci` linter with default settings
- Added concurrency to the `pizza bake` command
- Updated `pizza insights contributors` to use https://github.com/charmbracelet/bubbles#table for outputting results

Until https://github.com/open-sauced/go-api/issues/2 is resolved, `pizza bake` commands will fail with `400 Bad Request`.

**Comment by cecobask** [comment]
The problems with [open-sauced/go-api](https://github.com/open-sauced/go-api/issues/2) were resolved.
I believe there are no more changes to be added here.
@jpmcb, can you please take a look when you get a chance?

File: main.go
Comment (cecobask) [inline main.go:20]: Removed this log entry since the `Execute` method already outputs the error.
Previously, the error was outputted twice.

File: cmd/root/root.go
Comment (cecobask) [inline cmd/root/root.go:42]: Every time a command returned an error, the `help` output was also printed, together with the error.
I think this is cleaner. The help output will still be shown in case the user entered invalid input for commands or passed `--help` flag.

File: cmd/repo-query/repo-query.go
Comment (jpmcb) [inline cmd/repo-query/repo-query.go:74]: nit: Spacing was for readability

File: cmd/repo-query/repo-query.go
Comment (jpmcb) [inline cmd/repo-query/repo-query.go:79]: Looks like it's getting used in `contributors.go` which is why it was changed from a method on the `repoQueryAgent` struct to a publicly facing package function.

This is a pretty general utility function, so we should abstract it out into some a utility package in `pkg/`. Maybe something like `pkg/github` since it's pretty specific to GitHub

File: cmd/root/root.go
Comment (jpmcb) [inline cmd/root/root.go:17]: What's this empty constant block for?

File: cmd/root/root.go
Comment (jpmcb) [reply cmd/root/root.go:42]: > Every time a command returned an error, the help output was also printed

We should still dump the help message when an error is returned since it's usually due to some misconfiguration of flags or values. The standard approach in Cobra CLIs is to print the help message along with the error so the user can have a quick loop to correcting it.

File: pkg/api/constants.go
Comment (jpmcb) [inline pkg/api/constants.go:15]: Thanks for adding these!! 👍🏼

File: pkg/api/constants.go
Comment (jpmcb) [inline pkg/api/constants.go:31]: I'm generally not a fan of this approach. It always seems redundant. Why not just have callers get the endpoints directly through imports?

```go
myEndpoint := api.BetaEndpointURL
```

and I'm abit confused why we would need this? Generally, if a user provides `--endpoint`, we should take that whole sale and not check it against our keying of endpoints hard coded into the CLI. I may deploy my own API behind a local network and I want to provide that explicitly like:

```
pizza --endpoint https://api.opensauced.my-local-network.homelab ...
```

File: pkg/utils/root.go
Comment (jpmcb) [inline pkg/utils/root.go:9]: nit: This spacing was for readability

File: pkg/api/client.go
Comment (jpmcb) [inline pkg/api/client.go:37]: Why return the context? The background context should be handled by the API if for some reason it had a panic or exited early. And our control flow shouldn't be handling the background context anyways.

I wonder if it'd be better to do `context.TODO` since we're not really handling different context's control flow right now.

File: cmd/insights/contributors.go
Comment (jpmcb) [inline cmd/insights/contributors.go:26]: nit: `FilePath` field makes more sense to me.

File: cmd/insights/contributors.go
Comment (jpmcb) [inline cmd/insights/contributors.go:58]: Same as in the bake command, I believe these flags being mutually exclusive is confusing. Instead, I would like to see:

```
pizza insights contributors URL
```
```
pizza insights contributors URL --file my-other-repos.yaml
```
```
pizza contributors insights --file my-repos.yaml
```

File: cmd/insights/contributors.go
Comment (jpmcb) [inline cmd/insights/contributors.go:90]: This looks like a similar structuring as the bake command refactor you have: I don't think having the `waitgroup.Wait()` in a separate thread make sense. The main through should just wait for these all to finish.

And similar comments to the `errorChan`: these will get squashed and only the first error will surface and be returned.

File: cmd/insights/contributors.go
Comment (jpmcb) [inline cmd/insights/contributors.go:149]: Having a structured error here seems unnecessary. Typically I've only seen these to be useful when different error types need to be compared across callers of APIs. Like when using `stat` and `os`:
```go
_, err := os.Stat(path)
if os.IsNotExist(err) { 
    // handle the file at path not existin
}
```
but it doesn't seem we're doing that here. Let's remove this struct type in favor of traditional stacked error handling in Go.

File: cmd/insights/contributors.go
Comment (jpmcb) [inline cmd/insights/contributors.go:152]: `GetOwnerAndRepoFromURL` should be abstracted out of the `repoquery` package.

File: cmd/insights/contributors.go
Comment (jpmcb) [inline cmd/insights/contributors.go:239]: Same here: only the first error gets popped off the error channel. The others get squashed.

File: cmd/insights/contributors.go
Comment (jpmcb) [inline cmd/insights/contributors.go:185]: I think I'd prefer if we explicitly added `1` each time for each `go func` block. Otherwies, if another gets added, it won't be clear to other contributors what's going on here.

**Review by jpmcb — CHANGES_REQUESTED** [review]

File: pkg/api/constants.go
Comment (cecobask) [reply pkg/api/constants.go]: For some reason, I assumed the go api client can only be used with the server configurations that it originally has.
I've refactored it to allow the use of any endpoint url.

**Review by cecobask — COMMENTED** [review]
Thank you for the feedback, @jpmcb!
I've updated the code based on your suggestions.
Please, review at your convenience 👍🏻

**Comment by jpmcb** [comment]
> Thank you for the feedback, @jpmcb!

Happy to provide the feedback! This is looking good, great work. Huge lift. This is on my docket to dive back into today!

File: pkg/utils/arguments.go
Comment (jpmcb) [inline pkg/utils/arguments.go:10]: I think a `map[string]struct{}` may be slightly more efficient since it essentially creates an empty member on the right side which effectively implements a "set". But that's more a suggestion: take it or leave it!

File: cmd/insights/contributors.go
Comment (jpmcb) [inline cmd/insights/contributors.go:100]: Generally speaking, and this can be an additional issue we work on after this merges, we should also always provide a `-o / --output` flag that can be either "yaml" or "json" since we want people to be able to compose their data into other tools (like jq, grep, sed, awk, etc.)

**Review by jpmcb — COMMENTED** [review]
This is getting very close!!!! Nice work!!!

File: cmd/insights/contributors.go
Comment (cecobask) [reply cmd/insights/contributors.go:100]: I understand, I can refactor this to allow such an option as part of the work for https://github.com/open-sauced/pizza-cli/issues/10 👍🏻

**Comment by cecobask** [comment]
> This is getting very close!!!! Nice work!!!

Thanks, @jpmcb! I've updated the branch with the latest changes 👍🏻 
Let me know if something else needs to be done here.

**Review by jpmcb — APPROVED** [review]
Great work!! Just one small nit on a regression I noticed in using a custom endpoint for the pizza service (which I typically do for testing local changes). This might be an issue with how the client generates it's endpoints. Otherwise, let's get this in! 💪🏼

File: cmd/bake/bake.go
Comment (cecobask) [reply cmd/bake/bake.go:60]: Hi @jpmcb, I think it works as expected.
I haven't set up the pizza service locally but the output I'm getting indicates that the go client is using the correct URL.
This is the error output I'm getting when I use a custom URL:
```
Error: error while calling 'PizzaOvenServiceAPI.BakeARepositoryWithThePizzaOvenMicroservice' with repository "https://github.com/jpmcb/gopherlogs": Post "http://localhost:8080/v1/bake": dial tcp [::1]:8080: connect: connection refused
```
I think you're getting a 404 error because the repository might not be indexed yet, but I might be wrong.

**Review by bdougie — APPROVED** [review]
I have been watching this quietly in the background and can say it works as expected for me. 

re: localhost. It might be worth setting up some follow up issues to identify things we can fix and enhance after this is merged.

Something top of mind is making these contributor insights actionable with some context on  % and story telling.

Perhaps we get @isabensusan to assist in adding this to the design backlog to assist in representing this in the UI.

File: cmd/bake/bake.go
Comment (jpmcb) [reply cmd/bake/bake.go:60]: Good insight. I believe this is because it's attempting to use a `v1` endpoint. So, when I provide `http://localhost:8080`, it's calling into `http://localhost:8080/v1/bake` which is the actual API endpoint, not the local pizza endpoint I have setup. It's doing the right thing since the client is reaching into the routes it knows about. This is good for now unless we have a different local dev story we want to explore 👍🏼

**Pull request merged**
